lib/sources.nix

```
@@ -254,6 +254,64 @@ let
       outPath = builtins.path { inherit filter name; path = origSrc; };
     };
 
+  # urlToName : (URL | Path | String) -> String
+  #
+  # Transform a URL (or path, or string) into a clean package name.
+  urlToName = url_:
+    let
+      inherit (builtins) stringLength;
+      url = toString url_;
+      base = baseNameOf (lib.removeSuffix "/" (lib.last (lib.splitString ":" url)));
+      # chop away one git or archive-related extension
+      removeExt = name: let
+          matchExt = builtins.match "(.*)\\.(git|tar|zip|gz|tgz|bz|tbz|bz2|tbz2|lzma|txz|xz|zstd)" name;
+        in if matchExt != null then lib.head matchExt else name;
+      # apply function f to string x while the result shrinks
+      shrink = f: x: let v = f x; in if stringLength v < stringLength x then shrink f v else x;
+    in shrink removeExt base;
+
+  # shortRev : (String | Integer) -> String
+  #
+  # Given a package revision (like "refs/tags/v12.0"), produce a short revision ("12.0").
+  shortRev = rev_:
+    let
+      rev = toString rev_;
+      baseRev = baseNameOf rev;
```

No reason to do this other way:
```suggestion
      baseRev = baseNameOf (toString rev);
```